### DISCUSSION TITLE
HTTPS server with Colyseus instead of SSL Termination

### DISCUSSION THREAD

**Discussion by iratefun** [colyseus/colyseus#810] [Q&A] [answered]
Reading the Colyseus documentation, I saw that the **options.server** parameter is deprecated and the recommendation for SSL/TLS encryption is to use NGINX for encryption (SSL/TLS termination).  

The issue is that in a lot of secure environments it is simply prohibited via policy to run any sort of webserver, even internally, without encryption.  In order to get around this I'd have to utilize NGINX locally on the machine, and then proxy connections to ports on localhost, and overall it's just pain and additional maintenance overhead compared to just using https.createServer.

My question is, would it be possible to convince you guys to continue to allow us to use **options.server**, or perhaps even allow us to provide certificates to colyseus so that it creates a HTTPS server instead of only a HTTP one?  It would help me and my development team greatly to have this additional flexibility available.

**Comment by endel** [comment, marked as answer]
Hi @iratefun, welcome! 

You may still provide your own `server` option, but now from the `Transport` layer. There are 2 WebSocket transports available by default: using `ws` or `uWebSockets.js`: https://docs.colyseus.io/server/transport/#default-websocket-transport-via-ws

If you'd like to provide SSL/TLS certificates directly on the Node.js layer you could do so by providing options for the `Transport` implementation. 

I'm not sure if it's possible to reload the certificates without restarting the Node.js processes this way, though. When using NGINX I know this is possible. Would appreciate if you have any idea about this!

Kind regards,
Endel

**Reply by iratefun** [reply]
Thank you, i'll give it a go on my end utilizing Transport.  Regarding reloading certificates, that's not in my use-case as I simply restart my nodejs server processes on change/deployment.  

Since you're the developer, I have another question that you may be able to help with.  When using schemas, I noticed that basic types (Numbers Strings) etc are available, but not complex objects.  That's fine, as I can stringify JSON and use that to emulate a complex type by simply having a client reparse the data on a server state change.  My question is, is this a reasonable approach within your design, or is there another way you'd recommend me to approach this concept?

Also, I just want to commend you and say this software seems like a great project, not just for gaming.  The ability to share a server state using deltas among various connected clients is also very useful for a wide plethora of team oriented software.

**Reply by endel** [reply]
Thanks for the kind words @iratefun 💜 

The schemas can be composed of both primitive and "complex" types, which include other `Schema` instances, Arrays and Maps. As opposed to React's state management, the state in Colyseus is **mutable** - it's only possible to track what has changed by mutating properties in-place, instead of replacing the instances completely.

There's a page dedicated to schema definition here: https://docs.colyseus.io/state/schema/

_(The upcoming version `0.16` is almost stable, and you may be able to find more info on the new docs page 🙌  https://deploy-preview-163--colyseus-docs.netlify.app/examples)_

**Reply by iratefun** [reply]
Just want to say thank you @endel, utilizing transport allowed me to use my own https server.  Everything appears to be working great on my end.  I had to do a bit of mashing of example code, and some experimentation, but it works without issue.

May I recommend that you include a similar use case tutorial in your main docs?  I'm sure there are a lot of people with existing servers that would love to have an easy to follow basic room/gameserver/auth example which uses a http(s) server they've already implemented/built.  Just an example using self-signed certs and a http(s).createServer would suffice for most use cases I think.

Thanks again.
